Add workspace clone bits to frontend

sanjayankur31 opened the pull request:
- Adds service, updates reducer etc.
- Adds option to the context menu, which is shown in the main homepage workspace list, and in the resources side bar

@filippomc : I've added a TODO in the workspacepage.tsx file. Could you take a look? Is that Accordion unused?

filippomc commented on applications/osb-portal/src/components/workspace/drawer/WorkspaceDrawer.tsx:
```diff
@@ -124,7 +125,7 @@ export const WorkspaceDrawer: React.FunctionComponent<WorkspaceDrawerProps> = ({
       >
         <div className={`${open ? classes.drawerContent : ''} verticalFit`}>
-          <WorkspaceInteractions open={open} openResource={setCurrentResource} />
+          <WorkspaceInteractions open={open} openResource={setCurrentResource} cloneWorkspace={cloneWorkspace}/>
```
The cloneWorkspace action would be better passed implicitly by the redux connector as other callbacks are here. If you follow the path from index.ts you will find the redux connector logic and you'll also see tha WorkspaceActions are passed at the end of the day to the WorkspaceInteractions component, so no need for propagating properties

filippomc commented on applications/osb-portal/src/components/workspace/drawer/WorkspaceDrawer.tsx:
```diff
@@ -89,9 +89,10 @@ interface WorkspaceDrawerProps {
 }
 
-export const WorkspaceDrawer: React.FunctionComponent<WorkspaceDrawerProps> = ({ user, children, workspace }) => {
+export const WorkspaceDrawer: React.FunctionComponent<WorkspaceDrawerProps> = ({ user, children, workspace, cloneWorkspace }) => {
```
See comment below: it should not be needed to pass the cloneWorskpace action explicitly here

filippomc commented on applications/osb-portal/src/components/workspace/drawer/WorkspaceInteractions.tsx:
```diff
@@ -192,7 +193,7 @@ export default (props: WorkspaceProps | any) => {
 
             <Box p={2}>
-              <WorkspaceActionsMenu workspace={workspace} user={props.user} updateWorkspace={props.updateWorkspace} deleteWorkspace={props.deleteWorkspace} refreshWorkspaces={props.refreshWorkspace} />
+              <WorkspaceActionsMenu workspace={workspace} user={props.user} updateWorkspace={props.updateWorkspace} deleteWorkspace={props.deleteWorkspace} refreshWorkspaces={props.refreshWorkspace} cloneWorkspace={props.cloneWorkspace} />
```
Just for completeness: here we could have done the redux props mapping but wasn't done from the beginning. No choice is correct of wrong by itself, but I think it's good to keep consistency by having either having the component connected to redux or not, and if it is, using the connector to bind the actions. as props

sanjayankur31 replied:
Since this component is not connected to the redux store yet, I've left this as is for the moment.

filippomc commented on applications/osb-portal/src/pages/WorkspaceOpenPage.tsx:
```diff
@@ -18,7 +18,7 @@ export default (props: any) => {
     <Box className="verticalFill">
       <Divider variant="fullWidth" />
-      <WorkspaceDrawer app={app} />
+      <WorkspaceDrawer app={app} cloneWorkspace={props.cloneWorkspace}/>
```
same here

filippomc commented on applications/osb-portal/src/pages/WorkspacePage.tsx:
```diff
@@ -212,15 +216,25 @@ export const WorkspacePage = (props: any) => {
             </AccordionSummary>
             <AccordionDetails>
-              <WorkspaceInteractions workspace={workspace} open={true} user={props.user} refreshWorkspace={() => { setRefresh(!refresh) }} openResource={handleResourceClick} />
+              <WorkspaceInteractions workspace={workspace} open={true} user={props.user} refreshWorkspace={() => { setRefresh(!refresh) }} openResource={handleResourceClick} cloneWorkspace={props.cloneWorkspace}/>
```
same here

filippomc commented on applications/osb-portal/src/store/actions/workspaces.ts:
```diff
@@ -1,6 +1,6 @@
 import { WorkspaceActions } from '../reducers/workspaces'
 
-export const { selectWorkspace, refreshWorkspace, refreshWorkspaces, deleteWorkspace, updateWorkspace, resourceAdded } = WorkspaceActions;
+export const { selectWorkspace, refreshWorkspace, refreshWorkspaces, deleteWorkspace, updateWorkspace, resourceAdded, cloneWorkspace } = WorkspaceActions;
```
as this was added, WorkspaceInteraction will have the cloneWorkspace prop added without passing it explicitly

filippomc reviewed (commented):
I added some comments about the passing of the new `cloneWorskpace` action.
Also this new action is not really needed as the logic of the clone is the same of the one for new workspace creation, so the action `refreshWorkspaces` could have been used directly (in fact, the reducer does not do anything specific for cloneWorkspace).

I'm not requesting changes as what I see is not technically wrong, but still it would be nice to change to reduce entropy and have cleaner code

filippomc commented on applications/osb-portal/src/pages/WorkspacePage.tsx:
```diff
@@ -212,15 +216,25 @@ export const WorkspacePage = (props: any) => {
           </Box>
+          {
+            /*
+              TODO: is this accordion being used? Looks like a duplicate of the left resource panel below
```
@sanjayankur31 This accordion is used as a responsive alternative for small displays. You can see it by reducing the screen size

sanjayankur31 commented:
Thanks for the comments, I'lll tweak bits and update the PR---probably on Monday, though.

sanjayankur31 commented:
@filippomc : I've removed the unnecessary propagation of props now. Does that look OK?

About the actions---I think I understand what you're saying, but I'm not entirely clear on how to use the `refreshWorkspaces` action, for example. Could you explain it  a little please?

Oh, and I've left the `WorkspaceActionMenu` component with props for the time being, because that was the initial pattern used there. I've added a TODO there to remind us to update that to use redux also.

filippomc commented on applications/osb-portal/src/pages/WorkspacePage.tsx:
```diff
@@ -231,8 +245,13 @@ export const WorkspacePage = (props: any) => {
+          }
         <Box className={`verticalFit ${classes.workspaceResourcesInformation}`} display="flex" flexDirection="row">
-          <WorkspaceInteractions workspace={workspace} open={true} user={props.user} refreshWorkspace={() => { setRefresh(!refresh) }} openResource={handleResourceClick} />
+          <WorkspaceInteractions workspace={workspace} open={true} user={props.user} refreshWorkspace={() => { setRefresh(!refresh) }} openResource={handleResourceClick}  cloneWorkspace={props.cloneWorkspace}/>
```
no need to pass it here actually. But you were right to put it here in the beginning as WorkspaceInteractions was not imported correctly from the component/index.ts file here. I fixed that today with the delete issue

filippomc reviewed (commented):
Added a comment as there was a change that fixed the reason why you needed the new action to be passed explicitly to begin with.

To make everything even cleaner, I suggest to remove the `cloneWorkspace` action completely, as there is no logic attached to it, neither in the reducer nor in the middleware.
The workspace can be simply cloned and then the `refreshWorskpaces` action can be used to refresh the workspace list after the clone.

There is also a conflict to fix after my last change, sorry

sanjayankur31 commented:
Ah, awesome. Thanks. I'll make the changes and push new commits!

sanjayankur31 commented:
How does this look @filippomc ?

Also fixed a bug (regression) where the page (view) wasn't updated after the workspace was edited.

filippomc reviewed (approved):
Nice!
